cmd/config.go

```diff
@@ -309,6 +301,82 @@ type CAADistributedResolverConfig struct {
 	Proxies     []string
 }
 
+// TemporalLogDescription describes a single shard of a temporally shared
+// CT log
+type TemporalLogDescription struct {
+	URI         string
+	Key         string
+	WindowStart time.Time
+	WindowEnd   time.Time
+}
+
+// TemporalSet contains a set of temporal shards of a single log
+type TemporalSet struct {
+	Name   string
+	Shards []TemporalLogDescription
+}
+
+// Setup initializes the TemporalSet by parsing the start and end dates
+// and verifying WindowEnd > WindowStart
+func (ts *TemporalSet) Setup() error {
+	if ts.Name == "" {
+		return errors.New("Name cannot be empty")
+	}
+	if len(ts.Shards) == 0 {
+		return errors.New("temporal set contains no shards")
+	}
+	for i := range ts.Shards {
+		if ts.Shards[i].WindowEnd.Before(ts.Shards[i].WindowStart) || ts.Shards[i].WindowEnd.Equal(ts.Shards[i].WindowStart) {
+			return errors.New("WindowStart must be before WindowEnd")
+		}
+	}
+	return nil
+}
+
+// pick chooses the correct shard from a TemporalSet to use for the given
+// expiration time. In the case where two shards have overlapping windows
+// the earlier of the two shards will be chosen.
+func (ts *TemporalSet) pick(exp time.Time) (*TemporalLogDescription, error) {
+	for _, shard := range ts.Shards {
+		if exp.Before(shard.WindowStart) {
+			continue
+		}
+		if !exp.Before(shard.WindowEnd) {
+			continue
+		}
+		return &shard, nil
+	}
+	return nil, fmt.Errorf("no valid shard available for temporal set %q for expiration date %q", ts.Name, exp)
+}
+
+// Info returns the URI and key of the earliest valid shard
+func (ts *TemporalSet) Info(exp time.Time) (string, string, error) {
+	shard, err := ts.pick(exp)
+	if err != nil {
+		return "", "", err
+	}
+	return shard.URI, shard.Key, nil
+}
+
+// LogDescription contains the information needed to submit certificates
+// to a CT log and verify returned receipts
+type LogDescription struct {
+	URI             string
```

Comment:
Please add a comment: If TemporalSet is non-nil, then URI and Key should be absent.